internal/service/dynamodb/global_secondary_index.go

```diff
@@ -0,0 +1,710 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package dynamodb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
+	awstypes "github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+	"github.com/dlclark/regexp2"
+	"github.com/hashicorp/aws-sdk-go-base/v2/tfawserr"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-aws/internal/conns"
+	"github.com/hashicorp/terraform-provider-aws/internal/create"
+	"github.com/hashicorp/terraform-provider-aws/internal/enum"
+	"github.com/hashicorp/terraform-provider-aws/internal/errs"
+	"github.com/hashicorp/terraform-provider-aws/internal/flex"
+	"github.com/hashicorp/terraform-provider-aws/internal/slices"
+	"github.com/hashicorp/terraform-provider-aws/names"
+)
+
+const (
+	resNameGlobalSecondaryIndex = "global_secondary_index"
+)
+
+// @SDKResource("aws_dynamodb_global_secondary_index", name="Global Secondary Index")
+// @Testing(existsType="github.com/aws/aws-sdk-go-v2/service/dynamodb/types;types.TableDescription")
+func resourceGlobalSecondaryIndex() *schema.Resource {
+	//lintignore:R011
+	return &schema.Resource{
+		CreateWithoutTimeout: resourceGlobalSecondaryIndexCreate,
+		ReadWithoutTimeout:   resourceGlobalSecondaryIndexRead,
+		UpdateWithoutTimeout: resourceGlobalSecondaryIndexUpdate,
+		DeleteWithoutTimeout: resourceGlobalSecondaryIndexDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(createTableTimeout),
+			Delete: schema.DefaultTimeout(deleteTableTimeout),
+			Update: schema.DefaultTimeout(updateTableTimeoutTotal),
+		},
+
+		SchemaVersion: 1,
+
+		Schema: map[string]*schema.Schema{
+			names.AttrIndexARN: {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"table": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"hash_key": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```

Comment:
Rather than separate `hash_key` and `range_key` attributes, this should match the underlying API with `attribute_name` and `key_type`, where `key_type` can be one of `HASH` or `RANGE`